src/Language/Fixpoint/Smt/Theories.hs

```diff
@@ -371,81 +376,81 @@ interpSymbols =
     bvBopSort  = FFunc bitVecSort $ FFunc bitVecSort bitVecSort
 
 
-interpSym :: Symbol -> Raw -> Sort -> (Symbol, TheorySymbol)
+interpSym :: Symbol s -> Raw -> Sort s -> (Symbol s, TheorySymbol s)
 interpSym x n t = (x, Thy x n t Theory)
 
 maxLamArg :: Int
 maxLamArg = 7
 
-axiomLiterals :: [(Symbol, Sort)] -> [Expr]
-axiomLiterals lts = catMaybes [ lenAxiom l <$> litLen l | (l, t) <- lts, isString t ]
+axiomLiterals :: forall s. (Eq s) =>  [(Symbol s, Sort s)] -> [Expr s]
+axiomLiterals lts = catMaybes [ lenAxiom l <$> litLen (symbol  l) | (l, t) <- lts, isString t ]
   where
-    lenAxiom l n  = EEq (EApp (expr (strLen :: Symbol)) (expr l)) (expr n `ECst` intSort)
+    lenAxiom l n  = EEq (EApp (expr (strLen :: Symbol s)) (expr l)) (expr n `ECst` intSort)
     litLen        = fmap (Data.Text.length .  symbolText) . unLitSymbol
 
 --------------------------------------------------------------------------------
 -- | Constructors, Selectors and Tests from 'DataDecl'arations.
 --------------------------------------------------------------------------------
-dataDeclSymbols :: DataDecl -> [(Symbol, TheorySymbol)]
+dataDeclSymbols :: (Eq s) => DataDecl s -> [(Symbol s, TheorySymbol s)]
 dataDeclSymbols d = ctorSymbols d ++ testSymbols d ++ selectSymbols d
 
 -- | 'selfSort d' returns the _self-sort_ of 'd' :: 'DataDecl'.
 --   See [NOTE:DataDecl] for details.
 
-selfSort :: DataDecl -> Sort
+selfSort :: (Eq s) => DataDecl s -> Sort s
 selfSort (DDecl c n _) = fAppTC c (FVar <$> [0..(n-1)])
 
 -- | 'fldSort d t' returns the _real-sort_ of 'd' if 't' is the _self-sort_
 --   and otherwise returns 't'. See [NOTE:DataDecl] for details.
 
-fldSort :: DataDecl -> Sort -> Sort
+fldSort :: (Eq s) => DataDecl s -> Sort s -> Sort s
 fldSort d (FTC c)
   | c == ddTyCon d = selfSort d
 fldSort _ s        = s
 
 --------------------------------------------------------------------------------
-ctorSymbols :: DataDecl -> [(Symbol, TheorySymbol)]
+ctorSymbols :: (Eq s) => DataDecl s -> [(Symbol s, TheorySymbol s)]
 --------------------------------------------------------------------------------
 ctorSymbols d = ctorSort d <$> ddCtors d
 
-ctorSort :: DataDecl -> DataCtor -> (Symbol, TheorySymbol)
-ctorSort d ctor = (x, Thy x (symbolRaw x) t Ctor)
+ctorSort :: (Eq s) => DataDecl s -> DataCtor s -> (Symbol s, TheorySymbol s)
+ctorSort d ctor = (FS x, Thy (FS x) (symbolRaw (FS x)) t Ctor)
   where
     x           = symbol ctor
```

Comment:
Do we want to consider changing `symbol :: a -> FixSymbol` to `symbol :: a -> Symbol s`? Then we don't lose the `s` information.

Reply:
Applies to the rest of this file.

Reply:
Yes, that's actually one thing I wanted to include in my summary. In fact, there are many functions in Names.hs which still return a FixSymbol. Those are more difficult to handle since we have to make the decision whether those operations make sense to the Abstract Symbol. By changing the return type from FixSymbol -> Symbol s, we could either ignore the Abstract symbol and change only the fix symbol, or we could define a typeclass which allows the Abstract symbol to perform the same operation as well.